### PR TITLE
nx-cugraph: add `ancestors` and `descendants`

### PR DESCRIPTION
**Pull request by eriknw**

### CODE REVIEW

File: python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py
Comment (rlratzel) [inline python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py:129]: Why are we keeping the commented-out option 2 code?

File: python/nx-cugraph/nx_cugraph/classes/graph.py
Comment (rlratzel) [inline python/nx-cugraph/nx_cugraph/classes/graph.py:476]: I'm curious about this pattern here; why is there a separate `_neighbors` method instead of just including it inline here in `neighbors`?

File: python/nx-cugraph/nx_cugraph/interface.py
Comment (rlratzel) [inline python/nx-cugraph/nx_cugraph/interface.py:249]: Is this reason enough to move this PR to Draft until there's some kind of caching mechanism added?

File: python/nx-cugraph/nx_cugraph/classes/graph.py
Comment (eriknw) [reply python/nx-cugraph/nx_cugraph/classes/graph.py:476]: Different return type. Sometimes we want the cupy array. `G.neighbors` is part of the networkx API and is an iterator.

File: python/nx-cugraph/nx_cugraph/interface.py
Comment (eriknw) [reply python/nx-cugraph/nx_cugraph/interface.py:249]: I don't think that's necessary. I think this PR is reviewable and mergable. We already have this of repeated conversions elsewhere. Perhaps we could create an issue to track this.

File: python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py
Comment (eriknw) [reply python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py:62]: No preference. I can mention it.

File: python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py
Comment (eriknw) [reply python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py]: Chose option 1 and removed option 2. Perhaps we can revisit option 2 once we make renumbering helper functions.

**Comment by eriknw** [comment]
@rlratzel what do you think about the three tests that fail due to running out of GPU memory? I don't think the graphs are particularly large. What can we try? Can we investigate? Could it help to call `gc.collect()` regularly during tests (since maybe device memory is being help by released Python objects, but Python's garbage collection doesn't trigger b/c host memory usage is low)? Or, should we just skip the tests (and maybe run them one-by-one)?

File: python/nx-cugraph/nx_cugraph/interface.py
Comment (rlratzel) [reply python/nx-cugraph/nx_cugraph/interface.py:249]: ok, I created this issue: https://github.com/rapidsai/cugraph/issues/4055

File: python/nx-cugraph/nx_cugraph/algorithms/dag.py
Comment (rlratzel) [reply python/nx-cugraph/nx_cugraph/algorithms/dag.py:30]: I was thinking about this more and realized it might be better to have a function each NX algo defined then called to verify it's inputs, then that same function could be called by backends too - no code duplication and guaranteed consistency.  This would also make it possible to support applications that call `nx-cugraph` directly.  The pattern could be something like this:
```python
######## NX dispatcher
def _dispatch(func):
    def _set_assert_args(aa_func):
        func._assert_args = aa_func
    func._arg_assertion = _set_assert_args
    ...
    return func

######## NX algo
@_dispatch
def foo(a, b):
  foo._assert_args(a, b)
  return a + b

@foo._arg_assertion
def _(a, b):
    if not(isinstance(a, int)):
        raise TypeError(f"expected int for a, got {type(a).__name__}")
    if not(isinstance(b, int)):
        raise TypeError(f"expected int for b, got {type(b).__name__}")

######## backend
def backend_foo(a, b):
    foo._assert_args(a, b)
    return a + b
```

but I agree that this would be a lot of work to refactor every dispatchable NX algo to apply this pattern everywhere.  We could do this incrementally though...
I don't think this should block this PR though, but maybe we should file an issue to NX (cc @MridulS)

Comment (rlratzel) [reply python/nx-cugraph/nx_cugraph/algorithms/dag.py:30]: I should say, I don't think the lack of reusable algo arg assertion funcs should block the PR, but I think the update to check if `source is not None` and raise `ValueError` to match NX should be done.

**Comment by rlratzel** [comment]
> @rlratzel what do you think about the three tests that fail due to running out of GPU memory? I don't think the graphs are particularly large. What can we try? Can we investigate? 

I was wondering if we have a memory leak, possibly related to PLC graph creation during the conversion from the NX Graph to the PLC Graph happening more often now that `nx-cugraph` has implemented more algos.  Do you see this when you run tests locally, or see anything that resembles a leak when polling `nvidia-smi` during testing?  

> Could it help to call `gc.collect()` regularly during tests (since maybe device memory is being help by released Python objects, but Python's garbage collection doesn't trigger b/c host memory usage is low)? 

Maybe? If this is happening during tests, I'd be concerned that it could happen in a large user application too, and they may not know or would prefer not to manually invoke the GC.  I'll try recreating this on my workstation using your branch.  I can then add a test setup hook that calls `gc.collect()` before each test to see if that makes a difference.

**Comment by rlratzel** [comment]
@eriknw I just ran your branch on my workstation and saw the same results - the same three tests failed with the same OOM error.  I was polling `nvidia-smi` and never saw it exceed ~800MB, and my workstation GPUs have (I think) more memory than the V100s in the CI runners (I have 50GB cards).

I don't think this is a leak or anything `gc.collect()` will fix; I think there's a bug somewhere in the algos (or combination of algos) called from the tests.  Are you seeing the same result when you run these tests locally?

I suspect something is causing an abrupt spike in memory which fails the test and the memory is reclaimed faster than my `nvidia-smi` polling interval, or there's a bug where we're computing an amount of memory to allocate that's much too big.  I'm just guessing at this point though.  I'll keep looking...

image

**Comment by rlratzel** [comment]
@eriknw - I have a minimal repro script.  Can you use this to debug further?  I'm going to stop for the night but I can help next week.
```python
import networkx as nx
from networkx.algorithms import flow

G = nx.fast_gnp_random_graph(100, 0.0575, seed=42)
cut = nx.minimum_node_cut(G, flow_func=flow.boykov_kolmogorov)

G = nx.DiGraph()
G.add_nodes_from([0, 1, 2])
list(nx.algorithms.dag.antichains(G))
```

```
(base) root@98e76e671240:/Projects/eriknw/cugraph# NETWORKX_AUTOMATIC_BACKENDS=cugraph python oom_repro.py
Traceback (most recent call last):
  File "/Projects/eriknw/cugraph/oom_repro.py", line 9, in <module>
    list(nx.algorithms.dag.antichains(G))
  File "/opt/conda/lib/python3.10/site-packages/networkx/algorithms/dag.py", line 940, in antichains
    TC = nx.transitive_closure_dag(G, topo_order)
  File "/opt/conda/lib/python3.10/site-packages/networkx/utils/decorators.py", line 770, in func
    return argmap._lazy_compile(__wrapper)(*args, **kwargs)
  File "<class 'networkx.utils.decorators.argmap'> compilation 28", line 4, in argmap_transitive_closure_dag_25
    import inspect
  File "/opt/conda/lib/python3.10/site-packages/networkx/utils/backends.py", line 576, in __call__
    return self.orig_func(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/networkx/algorithms/dag.py", line 811, in transitive_closure_dag
    TC.add_edges_from((v, u) for u in nx.descendants_at_distance(TC, v, 2))
  File "/opt/conda/lib/python3.10/site-packages/networkx/utils/backends.py", line 569, in __call__
    return self._convert_and_call(
  File "/opt/conda/lib/python3.10/site-packages/networkx/utils/backends.py", line 817, in _convert_and_call
    result = getattr(backend, self.name)(*converted_args, **converted_kwargs)
  File "/Projects/eriknw/cugraph/python/nx-cugraph/nx_cugraph/utils/decorators.py", line 86, in __call__
    return self.__wrapped__(*args, **kwargs)
  File "/Projects/eriknw/cugraph/python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py", line 239, in descendants_at_distance
    distances, predecessors, node_ids = plc.bfs(
  File "bfs.pyx", line 181, in pylibcugraph.bfs.bfs
  File "utils.pyx", line 53, in pylibcugraph.utils.assert_success
RuntimeError: non-success value returned from cugraph_bfs: CUGRAPH_UNKNOWN_ERROR std::bad_alloc: out_of_memory: CUDA error at: /opt/conda/include/rmm/mr/device/cuda_memory_resource.hpp
```

**Comment by eriknw** [comment]
> Do you see this when you run tests locally, or see anything that resembles a leak when polling nvidia-smi during testing?

Yeah, I see it (just like you) when I run the tests locally. However, if I run each failing test individually, they pass.

Thanks for investigating and making a reproducer!

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: python/nx-cugraph/nx_cugraph/algorithms/dag.py
Comment (eriknw) [reply python/nx-cugraph/nx_cugraph/algorithms/dag.py:30]: NetworkX raises `NetworkXError` for these algorithms if source is `None`, so we are correct how we match in this case. I think we may need to be more careful about `None` as a node (which isn't allowed) elsewhere--you're right, `ValueError: None cannot be a node` is somewhat common.

File: ci/test_python.sh
Comment (rlratzel) [inline ci/test_python.sh:118]: We can leave this in place in order to get this PR merged, but I think this is masking a bug (likely in pylibcugraph/libcugraph, not in nx-cugraph).

I filed [this issue](https://github.com/rapidsai/cugraph/issues/4067) and I'll open a PR to remove these lines and the skipped tests separately, which will be merged when the bug is fixed.

**Review by rlratzel — APPROVED** [review]

File: ci/test_python.sh
Comment (rlratzel) [reply ci/test_python.sh:118]: The PR to eventually fix the bug is #4068

**Pull request merged**
